crates/kernel/src/tasks/scheduler.rs

```diff
@@ -119,76 +160,70 @@ pub enum SchedulerError {
-    /// The final result of a task, to be sent back to the task's subscriber if there is one.
-    /// (Otherwise the result is thrown away).
-    Result(TaskId, TaskResult),
-    /// A request to fork a new task.
-    Fork {
-        fork_request: Fork,
-        reply: oneshot::Sender<TaskId>,
-        session: Arc<dyn Session>,
-    },
-    /// A request for a description of another task.
-    Describe(TaskId, oneshot::Sender<Vec<TaskDescription>>),
-    Kill {
-        requesting_task_id: TaskId,
-        victim_task_id: TaskId,
-        sender_permissions: Perms,
-        result_sender: oneshot::Sender<Var>,
-    },
-    Resume {
-        requesting_task_id: TaskId,
-        queued_task_id: TaskId,
-        sender_permissions: Perms,
-        return_value: Var,
-        result_sender: oneshot::Sender<Var>,
-    },
-    Disconnect(TaskId, Objid),
-    Retry(TaskId),
-}
-
-/// Public facing interface for the scheduler.
 impl Scheduler {
     pub fn new(database: Arc<dyn Database + Send + Sync>, config: Config) -> Self {
         let config = Arc::new(config);
         let (control_sender, control_receiver) = crossbeam_channel::unbounded();
-        let inner = Inner {
-            input_requests: Default::default(),
+        let inner = TaskQ {
             tasks: Default::default(),
+            suspended: Default::default(),
         };
         Self {
             running: Arc::new(AtomicBool::new(false)),
             database,
             next_task_id: Default::default(),
-            inner: RwLock::new(inner),
+            task_q: Arc::new(Mutex::new(inner)),
             config,
             control_sender,
             control_receiver,
         }
     }
 
     /// Execute the scheduler loop, run from the server process.
+    #[instrument(skip(self))]
     pub fn run(self: Arc<Self>) {
         self.running.store(true, Ordering::SeqCst);
-        self.clone().do_process();
+        info!("Starting scheduler loop");
+        loop {
+            let is_running = self.running.load(Ordering::SeqCst);
+            if !is_running {
+                warn!("Scheduler stopping");
+                break;
+            }
+            {
+                // Look for tasks that need to be woken (have hit their wakeup-time), and wake them.
+                let mut inner = self.task_q.lock().unwrap();
+                let now = Instant::now();
+
+                // We need to take the tasks that need waking out of the suspended list, and then
+                // rehydrate them.
+                let to_wake = inner
+                    .suspended
+                    .iter()
+                    .filter_map(|(task_id, sr)| match &sr.wake_condition {
+                        WakeCondition::Time(t) => (*t <= now).then_some(*task_id),
+                        _ => None,
+                    })
+                    .collect::<Vec<_>>();
+
+                for task_id in to_wake {
+                    let sr = inner.suspended.remove(&task_id).unwrap();
```

Comment:
Can't wait for [`extract_if`](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.extract_if) to be stabilized